pkg/elb/cleanup.go

```diff
@@ -0,0 +1,299 @@
+package elb
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/ec2/ec2iface"
+	"github.com/aws/aws-sdk-go/service/elb"
+	"github.com/aws/aws-sdk-go/service/elb/elbiface"
+	"github.com/aws/aws-sdk-go/service/elbv2"
+	"github.com/aws/aws-sdk-go/service/elbv2/elbv2iface"
+	"github.com/kris-nova/logger"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/pkg/cloudprovider"
+	awsprovider "k8s.io/kubernetes/pkg/cloudprovider/providers/aws"
+
+	api "github.com/weaveworks/eksctl/pkg/apis/eksctl.io/v1alpha5"
+	"github.com/weaveworks/eksctl/pkg/eks"
+)
+
+const (
+	loadBalancerKindClassic = iota
+	loadBalancerKindNetwork = iota
+)
+
+type loadBalancer struct {
+	kind                  int
+	ownedSecurityGroupIDs map[string]struct{}
+}
+
+// Cleanup finds and deletes any dangling ELBs associated to a Kubernetes Service
+func Cleanup(ctx context.Context, ec2API ec2iface.EC2API, elbAPI elbiface.ELBAPI, elbv2API elbv2iface.ELBV2API,
+	client *eks.Client, clusterConfig *api.ClusterConfig) error {
+
+	deadline, ok := ctx.Deadline()
+	if !ok {
+		return fmt.Errorf("not context deadline set in call to elb.Cleanup()")
+	}
+
+	kubernetesCS, err := client.NewClientSet()
+	if err != nil {
+		return err
+	}
+	services, err := kubernetesCS.CoreV1().Services(metav1.NamespaceAll).List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	// Delete Services of type 'LoadBalancer', collecting their ELBs to wait for them to be deleted later on
+	loadBalancers := map[string]loadBalancer{}
+	for _, s := range services.Items {
+		if s.Spec.Type == corev1.ServiceTypeLoadBalancer {
+			name := cloudprovider.DefaultLoadBalancerName(&s)
+			kind := getLoadBalancerKind(&s)
+			ctx, cleanup := context.WithTimeout(context.Background(), 30*time.Second)
+			securityGroupIDs, err := getSecurityGroupsOwnedByLoadBalancer(ctx, ec2API, elbAPI,
+				clusterConfig.Metadata.Name, name, kind)
+			cleanup()
```

Comment:
What is this cleanup function?

Reply:
It cancels the context, see https://golang.org/pkg/context/#WithTimeout :

> Canceling this context releases resources associated with it, so code should call cancel as soon as the operations running in this Context complete:
>```golang
>func slowOperationWithTimeout(ctx context.Context) (Result, error) {
>	  ctx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
>	  defer cancel()  // releases resources if slowOperation completes before timeout elapses
>	  return slowOperation(ctx)
>}
> ```

Reply:
@2opremio can we rename it to `cancelContext()` (in the next PR) so it doesn't get confused with the cleanup happening in this package?